### PR TITLE
Fix Issue #34923: Inferred dtype at the end of df explode method

### LINKED ISSUE #34923: Infer dtype when using `df.explode()`ENH:
#### Is your feature request related to a problem?

Yes. Currently, the `df.explode` method always returns an object for the column being exploded. This leads to loss of information about the `dtype` of the exploded column.

 E.g.
```
s = pd.Series([1,2,3]) # <- dtype('int64')
df = pd.DataFrame({'A': [s, s, s, s], 'B': 1})
df.explode("A").dtypes
```
|    | 0      |
|:---|:-------|
| A  | object |
| B  | int64  |

It would be great if pandas could return the underlying dtype if it was consistent across all rows. (Or return the best dtype (int -> float -> object).)

#### Describe the solution you'd like

1. *solution 1*: The best case scenario would be where pandas would directly infer the dtype if it was consistent (ignoring NaNs) across the across the row.
```
s = pd.Series([1,None,3]) # <- dtype('float64')
df = pd.DataFrame({'A': [s, s, s, s], 'B': 1}) # <- empty list is converted to NaN 
df.explode("A").dtypes
```
|    | 0      |
|:---|:-------|
| A  | float64 |
| B  | int64  |

2. *solution 2*: Providing a argument to force inferring the dtype:
```
s = pd.Series([1,None,3]) # <- dtype('float64')
df = pd.DataFrame({'A': [s, s, s, s], 'B': 1}) # <- empty list is converted to NaN 
df.explode("A", infer_type=True).dtypes
```
|    | 0      |
|:---|:-------|
| A  | float64 |
| B  | int64  |

#### Describe alternatives you've considered

Currently, I use the following workaround:
```
s = pd.Series([1,None,3]) # <- dtype('float64')
df = pd.DataFrame({'A': [s, s, s, s], 'B': 1}) # <- empty list is converted to NaN 

d = df.A[0].dtype
df2 = df.explode("A")
df2.A = df2.A.astype(d)
```

#### API breaking implications

Not sure.

### PR DESCRIPTION
**Pull request by SanthoshBala18**
- [ ] closes #34923 
- [ ] tests added / passed
          pandas/tests/frame/methods/test_explode.py, method test_inferred_dtype
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -318,6 +318,7 @@ Other enhancements
 - :meth:`DataFrame.cov` and :meth:`Series.cov` now support a new parameter ddof to support delta degrees of freedom as in the corresponding numpy methods (:issue:`34611`).
 - :meth:`DataFrame.to_html` and :meth:`DataFrame.to_string`'s ``col_space`` parameter now accepts a list or dict to change only some specific columns' width (:issue:`28917`).
 - :meth:`DataFrame.to_excel` can now also write OpenOffice spreadsheet (.ods) files (:issue:`27222`)
+- :meth:'DataFrame.explode' can now infer dtypes (:issue:`34923`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:321]: this goes in api breaking changes, can you expand on this slightly.

File: pandas/tests/series/methods/test_explode.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 def test_basic():
     s = pd.Series([[0, 1, 2], np.nan, [], (3, 4)], index=list("abcd"), name="foo")
-    result = s.explode()
+    result = s. explode()
```
Comment (jreback) [inline pandas/tests/series/methods/test_explode.py:10]: remove this space

File: pandas/tests/series/methods/test_explode.py [hunk 3]
```diff
@@ -118,3 +118,11 @@ def test_duplicate_index():
     result = s.explode()
     expected = pd.Series([1, 2, 3, 4], index=[0, 0, 0, 0], dtype=object)
     tm.assert_series_equal(result, expected)
+
+
+def test_inferred_dtype():
+    s = pd.Series([1, None, 3])
```
Comment (jreback) [inline pandas/tests/series/methods/test_explode.py:124]: add the issue number here

File: pandas/tests/series/methods/test_explode.py [hunk 4]
```diff
@@ -118,3 +118,11 @@ def test_duplicate_index():
...
+    df = pd.DataFrame({'A': [s, s, s, s], "B": 1})
+    result = df.explode("A").dtypes.values
+    expected = np.array([np.dtype('float64'), np.dtype('int64')])
```
Comment (jreback) [inline pandas/tests/series/methods/test_explode.py:127]: always use `assert_frame_equal`, numpy loses types

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
you may need to update the example in the docs as well

File: doc/source/whatsnew/v1.1.0.rst [hunk 5]
```diff
@@ -661,6 +661,38 @@ apply and applymap on ``DataFrame`` evaluates first row/column only once
...
+    In [3]: df.explode("A").dtypes
+    Out[3]:
+    A    object
+    B     int64
+    dtype: object
+
+*New behavior*:
+
+.. code-block:: ipython
+
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:689]: use an ipython block, so it runs the code for this part

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm pls merge master and ping on green

**Comment by SanthoshBala18** [comment]
@jreback, merged the master branch..

File: doc/source/whatsnew/v1.1.0.rst (hunk 5 above)
Comment (jreback) [reply doc/source/whatsnew/v1.1.0.rst:744]: you also need to remove the i python prompts 

they are added automatically

**Comment by SanthoshBala18** [comment]
@jreback Can we restart the checks.?

**Comment by simonjayhawkins** [comment]
I'm -1 on this. This is a breaking change and not a bug.

the change for lists of ints with some rows containing empty lists now being cast to float should wait till a major release and maybe use the nullable integer type instead.

current
```
>>> pd.__version__
'1.1.0rc0'
>>>
>>> s = pd.Series([[1, 2, 3, 4], [], [5, 6, 7]])
>>> s
0    [1, 2, 3, 4]
1              []
2       [5, 6, 7]
dtype: object
>>>
>>> res = s.explode()
>>> res
0      1
0      2
0      3
0      4
1    NaN
2      5
2      6
2      7
dtype: object
>>>
>>> type(res.iloc[1])
<class 'int'>
>>>
```

this PR
```
>>> pd.__version__
'1.1.0rc0+13.gf6144512d'
>>>
>>> s = pd.Series([[1, 2, 3, 4], [], [5, 6, 7]])
>>> s
0    [1, 2, 3, 4]
1              []
2       [5, 6, 7]
dtype: object
>>>
>>> res = s.explode()
>>> res
0    1.0
0    2.0
0    3.0
0    4.0
1    NaN
2    5.0
2    6.0
2    7.0
dtype: float64
>>>
>>> type(res.iloc[1])
<class 'numpy.float64'>
>>>

```

**Comment by jreback** [comment]
closing this. see my comments on the issue (basically need to add an option `downcast=` instead here.

**Pull request closed without merging**